### PR TITLE
MPS support

### LINKED ISSUE #1215: Support for MPS?
I was wondering if torchtune could support apple silicon. It looks like all the examples are for CUDA.

### PR DESCRIPTION
**Pull request by SalmanMohammadi**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

closes #1215

Please link to any issues this PR addresses.

#### Changelog

Added MPS support. Tested with `generate_v2`, `eleuther_eval`, `lora_finetune_single_device`

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: torchtune/generation/_generation.py [hunk 1]
```diff
@@ -325,6 +321,11 @@ def generate(
 
     # keeps track at a high level if we've already hit a stop token in a sequence so we can early stop
     stop_token_reached = torch.zeros(bsz, dtype=torch.bool, device=prompt.device)
+    stop_tokens = (
```
Comment (joecummings) [inline torchtune/generation/_generation.py:324]: Any reason?

File: torchtune/generation/_generation.py [hunk 2]
```diff
@@ -325,6 +321,11 @@ def generate(
...
+        torch.tensor(stop_tokens, device=prompt.device, dtype=tokens.dtype)
```
Comment (SalmanMohammadi) [inline torchtune/generation/_generation.py:325]: Not sure why but outputs from the model on my device were `int32`. The prompt is `int64` and `torch.isin` was complaining when `stop_tokens.dtype != tokens.dtype` later on. I wasn't ready for further investigation.

File: torchtune/generation/_generation.py (hunk 1 above)
Comment (SalmanMohammadi) [reply torchtune/generation/_generation.py:324]: see below

File: torchtune/generation/_generation.py (hunk 2 above)
Comment (joecummings) [reply torchtune/generation/_generation.py:325]: carry on

**Review by joecummings — APPROVED** [review]

File: torchtune/generation/_generation.py (hunk 2 above)
Comment (SalmanMohammadi) [reply torchtune/generation/_generation.py:325]:
```python
import torch
for dvc in ["cpu", "cuda"]:
    print(dvc)
    device = torch.device("cuda")
    x = torch.arange(10, device = device, dtype=torch.int64)
    y = torch.arange(10, device = device, dtype=torch.int32)
    torch.isin(x, y)
```
So fine on CPU/CUDA
BUT this seems like expected behaviour on MPS? There's actually a test for this in core
https://github.com/pytorch/pytorch/blob/e5a57932f06746a7f6779a97b4e11394b614fa81/test/test_mps.py#L8512

Comment (SalmanMohammadi) [reply torchtune/generation/_generation.py:325]: yeah there's a check for it in the MPS implementation https://github.com/pytorch/pytorch/blob/e5a57932f06746a7f6779a97b4e11394b614fa81/aten/src/ATen/native/mps/operations/TensorCompare.mm#L297

Comment (SalmanMohammadi) [reply torchtune/generation/_generation.py:325]: Maybe because the underlying MPS call doesn't support different datatypes?
https://developer.apple.com/documentation/metalperformanceshadersgraph/mpsgraph/equal(_:_:name:)?language=objc

Comment (SalmanMohammadi) [reply torchtune/generation/_generation.py:325]: This is probably it and there's NO WAY I'm loading up XCode to try it out. I hope you are contented.
https://xkcd.com/356/

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -55,6 +55,34 @@ Whenever running tests in torchtune, favor using the command line flags as much
 
 Note that the above flags can be combined with other pytest flags, so e.g. `pytest tests -m integration_test -k 'test_loss'` will run only recipe tests matching the substring `test_loss`.
 
+> [!NOTE]
+> For contributors who are developing on MPS, you may find that several of our tests fail due to precision differences falling outside of numerical tolerances on our tests. As of 27/09/2024 the following tests will fail for this reason:
+<details>
+  <summary>Failed Tests</summary>
```
Comment (ebsmothers) [inline CONTRIBUTING.md:61]: Commented in Discord, but why not just (a) add a new method in test_utils.py, (b) decorate each of these test cases with it, and (c) point to an issue tracking all of them in case someone wants to help debug any individual one?

File: torchtune/modules/kv_cache.py [hunk 4]
```diff
@@ -102,7 +102,7 @@ def update(
         k_out = self.k_cache
         v_out = self.v_cache
 
-        k_out.index_copy_(2, cache_pos, k_val)
-        v_out.index_copy_(2, cache_pos, v_val)
+        k_out[:, :, cache_pos] = k_val
+        v_out[:, :, cache_pos] = v_val
```
Comment (ebsmothers) [inline torchtune/modules/kv_cache.py:106]: Does this create an extra copy or anything? (Not a blocker, just curious)

**Review by ebsmothers — COMMENTED** [review]
Would be nice to have an actual gate on those failing unit tests with a pointer to a proper issue for now (otherwise it bloats up our CONTRIBUTING.md a lot). Other than that if you've run the generation recipe (yes just that one and not every single recipe), this looks good to me

File: torchtune/modules/kv_cache.py (hunk 4 above)
Comment (SalmanMohammadi) [reply torchtune/modules/kv_cache.py:106]: I'm not 100% sure. I drew inspiration from here https://github.com/huggingface/transformers/blob/2e24ee4dfa39cc0bc264b89edbccc373c8337086/src/transformers/cache_utils.py#L1168

**Review by ebsmothers — APPROVED** [review]
It’s beautiful
